### PR TITLE
Graph discipline2

### PR DESCRIPTION
**Pull request by o-**
This change enforces a more rigorous graph discipline. There is a distinction
between merge labels and branch labels. Merge labels can have an arbitrary
number of incomming 'gotos'. Branch labels can have exactly one incomming
source from a branch instruction.

To make checking the graph a local property branch labels must begin with '$'.

The graph discipline is strictly enforced in the check module.

In theory merge labels would not have a problem with fallthrough controlflow.
But to make the structure uniform fallthrough is forbidden too.

The change simplifies reasoning about instruction movements as can be seen
in the hoist_drop pass.

depends on #142  and obsoletes #138

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-OCAMLBUILD=ocamlbuild
+OCAMLBUILD=ocamlbuild -cflag -g
```
Comment (gasche) [inline Makefile:1]: this should not be necessary because the `true: debug` line in `_tags` should have the same effect. Why did you feel the need for it?

**Comment by gasche** [comment]
I like the idea.

In terms of the implementation I wonder if it would not be nicer to separate the two `label` instructions, because we could have

```ocaml
type merge_label = ML of raw_label
type branch_label = BL of raw_label

type instr =
| ...
| Merge_label of merge_label
| Branch_label of branch_label
| Goto of merge_label
| Branch of expr * branch_label * branch_label
| Osr of { label : merge_label; ... }
| ...
```

and you would end up with substantially less `label -> label_type` conversion noise.

(Of course we could consider merging what you have now, and making that transformation afterwards.)

File: Makefile (hunk 1 above)
Comment (o-) [reply Makefile:1]: ups, that was not meant to be in there. I still wonder if ocamldebug is really hideous or if I don't know how to use it.

**Comment by o-** [comment]
> In terms of the implementation I wonder if it would not be nicer to separate the two label instructions, because we could have

I initially wanted to go this way. The reason I went with this version was that I wanted to keep the 1:1 mapping between our source language instruction and instruction type.

Update: also there are a number of places where you do not care about the kind of label and there it would introduce a lot of noise. I suspect you would need to extend about 20 pattern matches that currently just match `| Label _`.

Basically either one is fine. If you find two instructions much better I can also still change it.

**Comment by gasche** [comment]
I propose to merge this now, and consider the change as a later PR, so that the diff gets evaluated on its own effect instead of mixing the current restructuring and the change-of-representation diffs.

File: Makefile (hunk 1 above)
Comment (mhyee) [reply Makefile:1]: Reminder: this PR got merged before the debug flags were removed.

File: check.ml [hunk 2]
```diff
@@ -127,6 +135,30 @@ let well_formed prog =
...
+        let _ = Instr.resolve instrs (MergeLabel l) in
+        check_fun_ref instr
+      | Branch (e, l1, l2) ->
+        let t1 = Instr.resolve instrs (BranchLabel l1) in
+        if (preds.(t1) <> [pc]) then raise (BranchLabelReused t1);
+        let t2 = Instr.resolve instrs (BranchLabel l2) in
+        if (preds.(t2) <> [pc]) then raise (BranchLabelReused t2);
+        if (t1 = t2) then raise (BranchLabelReused t1);
+        check_fun_ref instr
+      | Label (MergeLabel l| BranchLabel l) ->
```
Comment (mhyee) [inline check.ml:148]: Nitpick: missing space after `MergeLabel l`.

File: check.ml [hunk 3]
```diff
@@ -127,6 +135,30 @@ let well_formed prog =
...
+        if List.mem l !seen then raise (DuplicateLabel l);
+        seen := l :: !seen;
+        (* Entry point cannot be a label because if it where a branch
```
Comment (mhyee) [inline check.ml:151]: Typo: where -> were

File: check.ml [hunk 4]
```diff
@@ -127,6 +135,30 @@ let well_formed prog =
...
+         * target it would have two incomming control flows. *)
```
Comment (mhyee) [inline check.ml:152]: Typo: incomming -> incoming

File: disasm.ml [hunk 5]
```diff
@@ -58,8 +58,9 @@ let disassemble_instrs buf ?(format_pc = no_line_number) (prog : instructions) =
     | Drop var                        -> pr buf " drop %s" var
     | Assign (var, exp)               -> pr buf " %s <- %a" var dump_expr exp
     | Array_assign (var, index, exp)  -> pr buf " %s[%a] <- %a" var dump_expr index dump_expr exp
-    | Branch (exp, l1, l2)            -> pr buf " branch %a %s %s" dump_expr exp l1 l2
-    | Label label                     -> pr buf "%s:" label
+    | Branch (exp, l1, l2)            -> pr buf " branch %a $%s $%s" dump_expr exp l1 l2
+    | Label (MergeLabel label)        -> pr buf "%s:" label
+    | Label (BranchLabel label)        -> pr buf "$%s:" label
```
Comment (mhyee) [inline disasm.ml:63]: Nitpick: extra space screws up the alignment.

File: instr.ml [hunk 6]
```diff
@@ -55,13 +55,16 @@ and instruction =
...
   | Branch of expression * label * label
-  | Label of label
+  | Label of label_type
   | Goto of label
   | Print of expression
   | Assert of expression
   | Stop of expression
   | Osr of {label : label; cond : expression list; target : label position; map : osr_def list; }
   | Comment of string
+and label_type =
```
Comment (mhyee) [inline instr.ml]: These should probably be `Merge_label` and `Branch_label`.

**Review by mhyee — COMMENTED** [review]
I'll open a new issue so we can discuss splitting the label instruction.

(As a side note, I find

```
goto loop_body
$loop:
goto loop_body
loop_body:
```
a lot messier, which is a shame. But we've gone through enough pain from reusing labels.)
